src/transformers/modeling_utils.py

```diff
@@ -1574,7 +1574,15 @@ def _initialize_weights(self, module):
         """
         if getattr(module, "_is_hf_initialized", False):
             return
-        self._init_weights(module)
+
+        if is_deepspeed_zero3_enabled():
+            import deepspeed
+
+            with deepspeed.zero.GatheredParameters(list(module.parameters()), modifier_rank=0):
+                self._init_weights(module)
```

Comment:
This should only gather each module weights and not individual sub-module weights so LGTM.

Reply:
I moved the `GatheredParameters` context out of the recursion function. The context will be enabled only once.